File: tutorials/tutorial_08_mqtt_subscriber/cpp/src/main.cpp
#include <stdio.h>
#include <string.h>
#include <chrono>

#include <sl_iot/HubClient.hpp>
#include <csignal>
#include <ctime>

using namespace std;
using namespace sl;
using namespace sl_iot;
using json = sl_iot::json;


void onDataReceived(std::string topic, std::string message, TARGET target, void* arg)
{
    std::cout << "Message received !" << std::endl;
    json my_raw_data = json::parse(message);
    std::cout << "My received message : " << my_raw_data << std::endl;
    HubClient::sendLog("MQTT message received on topic " + topic,LOG_LEVEL::INFO); 
}


int main(int argc, char **argv) {

    STATUS_CODE status_iot;
    status_iot = HubClient::connect("mqtt_sub_app");
    if (status_iot != STATUS_CODE::SUCCESS) {
        std::cout << "Initialization error " << status_iot << std::endl;
        exit(EXIT_FAILURE);
    }

    // Topic to listen
    TARGET topic_prefix = TARGET::LOCAL_NETWORK;
    std::string topic_name = "/my_custom_data";

    HubClient::subscribeToMqttTopic(topic_name, onDataReceived, topic_prefix);

    // Main loop
    while (true) {

        sleep_ms(1000);
    }

    status_iot = HubClient::disconnect();
    if (status_iot != STATUS_CODE::SUCCESS) {
        std::cout << "Terminating error " << status_iot << std::endl;
        exit(EXIT_FAILURE);
    }

    return 0;
}
  

File: tutorials/tutorial_08_mqtt_subscriber/python/main.py
 
########################################################################
#
# Copyright (c) 2023, STEREOLABS.
#
# All rights reserved.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
########################################################################

import pyzed.sl_hub as hub
import time
import json


def on_data_received(topic, message, target):
    print("Message received !")
    my_raw_data = json.loads(message)
    print("My received message : ", my_raw_data)
    hub.HubClient.send_log("Message received on topic " + topic, hub.LOG_LEVEL.INFO)


def main():
    # Initialize the communication to ZED Hub, without a zed camera.
    status_hub = hub.HubClient.connect("sub_app")
    if status_hub != hub.STATUS_CODE.SUCCESS:
        print("Initialization error ", status_hub)
        exit(1)

    # Topic to listen to
    topic_name = "/my_custom_data"
    hub.HubClient.subscribe_to_topic(topic_name, on_data_received)

    # Main loop
    while True: 
        time.sleep(1)
    
    # Close the communication with ZED Hub properly.
    status_hub = hub.HubClient.disconnect()
    if status_hub != hub.STATUS_CODE.SUCCESS:
        print("Terminating error ", status_hub)
        exit(1)
    
    return


if __name__ == "__main__":
    main()
